agent/ls.go

```diff
@@ -0,0 +1,178 @@
+
+func (*agent) HandleLS(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	var query LSRequest
+	if !httpapi.Read(ctx, rw, r, &query) {
+		return
+	}
+
+	resp, err := listFiles(query)
+	if err != nil {
+		status := http.StatusInternalServerError
+		switch {
+		case errors.Is(err, os.ErrNotExist):
+			status = http.StatusNotFound
+		case errors.Is(err, os.ErrPermission):
+			status = http.StatusForbidden
+		default:
+		}
+		httpapi.Write(ctx, rw, status, codersdk.Response{
+			Message: err.Error(),
+		})
+		return
+	}
+
+	httpapi.Write(ctx, rw, http.StatusOK, resp)
+}
+
+func listFiles(query LSRequest) (LSResponse, error) {
+	var fullPath []string
+	switch query.Relativity {
+	case LSRelativityHome:
+		home, err := os.UserHomeDir()
+		if err != nil {
+			return LSResponse{}, xerrors.Errorf("failed to get user home directory: %w", err)
+		}
+		fullPath = []string{home}
+	case LSRelativityRoot:
+		if runtime.GOOS == "windows" {
+			if len(query.Path) == 0 {
+				return listDrives()
+			}
+			if !WindowsDriveRegex.MatchString(query.Path[0]) {
+				return LSResponse{}, xerrors.Errorf("invalid drive letter %q", query.Path[0])
+			}
+		} else {
+			fullPath = []string{"/"}
+		}
+	default:
+		return LSResponse{}, xerrors.Errorf("unsupported relativity type %q", query.Relativity)
+	}
+
+	fullPath = append(fullPath, query.Path...)
+	fullPathRelative := filepath.Join(fullPath...)
+	absolutePathString, err := filepath.Abs(fullPathRelative)
+	if err != nil {
+		return LSResponse{}, xerrors.Errorf("failed to get absolute path of %q: %w", fullPathRelative, err)
+	}
+
+	f, err := os.Open(absolutePathString)
+	if err != nil {
+		return LSResponse{}, xerrors.Errorf("failed to open directory %q: %w", absolutePathString, err)
+	}
+	defer f.Close()
+
+	stat, err := f.Stat()
+	if err != nil {
+		return LSResponse{}, xerrors.Errorf("failed to stat directory %q: %w", absolutePathString, err)
+	}
+
+	if !stat.IsDir() {
+		return LSResponse{}, xerrors.Errorf("path %q is not a directory", absolutePathString)
+	}
+
+	// `contents` may be partially populated even if the operation fails midway.
+	contents, _ := f.ReadDir(-1)
+	respContents := make([]LSFile, 0, len(contents))
+	for _, file := range contents {
+		respContents = append(respContents, LSFile{
+			Name:               file.Name(),
+			AbsolutePathString: filepath.Join(absolutePathString, file.Name()),
+			IsDir:              file.IsDir(),
+		})
+	}
+
+	absolutePath := pathToArray(absolutePathString)
+
+	return LSResponse{
+		AbsolutePath:       absolutePath,
+		AbsolutePathString: absolutePathString,
+		Contents:           respContents,
+	}, nil
+}
+
+func listDrives() (LSResponse, error) {
+	partitionStats, err := disk.Partitions(true)
+	if err != nil {
+		return LSResponse{}, xerrors.Errorf("failed to get partitions: %w", err)
+	}
+	contents := make([]LSFile, 0, len(partitionStats))
+	for _, a := range partitionStats {
+		name := a.Mountpoint + string(os.PathSeparator)
```

Comment:
This is inconsistent with the rest of the API, where we don't add a trailing path separator to directories.